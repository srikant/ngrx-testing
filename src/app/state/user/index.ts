import { createFeatureSelector, createSelector } from '@ngrx/store';
import { State } from '@state/user/user.reducer';
import * as fromUser from './user.reducer';

export const selectUserState = createFeatureSelector<State>('user');

export const selectUserIds = createSelector(
  selectUserState,
  fromUser.selectUserIds
);
export const selectUserEntities = createSelector(
  selectUserState,
  fromUser.selectUserEntities
);
export const selectAllUsers = createSelector(
  selectUserState,
  fromUser.selectAllUsers
);
export const selectUserTotal = createSelector(
  selectUserState,
  fromUser.selectUserTotal
);
export const selectSelectedUserId = createSelector(
  selectUserState,
  fromUser.getSelectedUserId
);
export const selectSelectedUser = createSelector(
  selectUserEntities,
  selectSelectedUserId,
  (entities, selectedUserId) => entities && entities[selectedUserId]
);
