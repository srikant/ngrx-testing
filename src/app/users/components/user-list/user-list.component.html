<ul>
  <li *ngFor="let user of users" trackBy="user.id">
    <a href="javascript:void(0)" (click)="selectUser.emit(user)">{{user.firstName}} {{user.lastName}}</a>
  </li>
</ul>