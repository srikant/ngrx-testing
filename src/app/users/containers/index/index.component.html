<button routerLink="add">Add User</button>
<app-user-list [users]="users | async" (selectUser)="onSelectUser($event)"></app-user-list>