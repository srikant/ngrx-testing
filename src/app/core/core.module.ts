import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import {
  ModuleWithProviders,
  NgModule,
  Optional,
  SkipSelf
} from '@angular/core';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { RootDbService } from './db/root.db.service';
import { UserService } from './services/user.service';

@NgModule({
  imports: [
    CommonModule,
    HttpClientModule,
    HttpClientInMemoryWebApiModule.forRoot(RootDbService, {
      delay: 500
    })
  ],
  declarations: [],
  providers: [UserService]
})
export class CoreModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CoreModule
    };
  }

  constructor(
    @Optional()
    @SkipSelf()
    parentModule?: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import it in the AppModule only'
      );
    }
  }
}
