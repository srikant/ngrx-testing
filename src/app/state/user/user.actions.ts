import { Update } from '@ngrx/entity';
import { Action } from '@ngrx/store';
import { User } from './user.model';

export enum UserActionTypes {
  LoadUser = '[User] Load User',
  LoadUserSuccess = '[User] Load User Success',
  LoadUserFail = '[User] Load User Fail',
  LoadUsers = '[User] Load Users',
  LoadUsersSuccess = '[User] Load Users Success',
  LoadUsersFail = '[User] Load Users Fail',
  AddUser = '[User] Add User',
  AddUserSuccess = '[User] Add User Success',
  AddUserFail = '[User] Add User Fail',
  AddUsers = '[User] Add Users',
  AddUsersSuccess = '[User] Add Users Success',
  AddUsersFail = '[User] Add Users Fail',
  UpdateUser = '[User] Update User',
  UpdateUserSuccess = '[User] Update User Success',
  UpdateUserFail = '[User] Update User Fail',
  UpdateUsers = '[User] Update Users',
  UpdateUsersSuccess = '[User] Update Users Success',
  UpdateUsersFail = '[User] Update Users Fail',
  DeleteUser = '[User] Delete User',
  DeleteUserSuccess = '[User] Delete User Success',
  DeleteUserFail = '[User] Delete User Fail',
  DeleteUsers = '[User] Delete Users',
  ClearUsers = '[User] Clear Users',
  SelectUser = '[User] Select User'
}

export class LoadUser implements Action {
  readonly type = UserActionTypes.LoadUser;

  constructor(public payload: { id: number }) {}
}

export class LoadUserSuccess implements Action {
  readonly type = UserActionTypes.LoadUserSuccess;

  constructor(public payload: { user: User }) {}
}

export class LoadUserFail implements Action {
  readonly type = UserActionTypes.LoadUserFail;

  constructor(public payload: { error: Error }) {}
}

export class LoadUsers implements Action {
  readonly type = UserActionTypes.LoadUsers;
}

export class LoadUsersSuccess implements Action {
  readonly type = UserActionTypes.LoadUsersSuccess;

  constructor(public payload: { users: User[] }) {}
}

export class LoadUsersFail implements Action {
  readonly type = UserActionTypes.LoadUsersFail;

  constructor(public payload: { error: Error }) {}
}

export class AddUser implements Action {
  readonly type = UserActionTypes.AddUser;

  constructor(public payload: { user: Partial<User> }) {}
}

export class AddUserSuccess implements Action {
  readonly type = UserActionTypes.AddUserSuccess;

  constructor(public payload: { user: User }) {}
}

export class AddUserFail implements Action {
  readonly type = UserActionTypes.AddUserFail;

  constructor(public payload: { error: Error }) {}
}

export class AddUsers implements Action {
  readonly type = UserActionTypes.AddUsers;

  constructor(public payload: { users: Partial<User>[] }) {}
}

export class AddUsersSuccess implements Action {
  readonly type = UserActionTypes.AddUsersSuccess;

  constructor(public payload: { users: User[] }) {}
}

export class AddUsersFail implements Action {
  readonly type = UserActionTypes.AddUsersFail;

  constructor(public payload: { error: Error }) {}
}

export class UpdateUser implements Action {
  readonly type = UserActionTypes.UpdateUser;

  constructor(public payload: { user: Partial<User> }) {}
}

export class UpdateUserSuccess implements Action {
  readonly type = UserActionTypes.UpdateUserSuccess;

  constructor(public payload: { update: Update<User> }) {}
}

export class UpdateUserFail implements Action {
  readonly type = UserActionTypes.UpdateUserFail;

  constructor(public payload: { error: Error }) {}
}

export class UpdateUsers implements Action {
  readonly type = UserActionTypes.UpdateUsers;

  constructor(public payload: { users: Partial<User>[] }) {}
}

export class UpdateUsersSuccess implements Action {
  readonly type = UserActionTypes.UpdateUsersSuccess;

  constructor(public payload: { update: Update<User>[] }) {}
}

export class UpdateUsersFail implements Action {
  readonly type = UserActionTypes.UpdateUsersFail;

  constructor(public payload: { error: Error }) {}
}

export class DeleteUser implements Action {
  readonly type = UserActionTypes.DeleteUser;

  constructor(public payload: { id: string }) {}
}

export class DeleteUserSuccess implements Action {
  readonly type = UserActionTypes.DeleteUserSuccess;

  constructor(public payload: { id: string }) {}
}

export class DeleteUserFail implements Action {
  readonly type = UserActionTypes.DeleteUserFail;

  constructor(public payload: { error: Error }) {}
}

export class DeleteUsers implements Action {
  readonly type = UserActionTypes.DeleteUsers;

  constructor(public payload: { ids: string[] }) {}
}

export class ClearUsers implements Action {
  readonly type = UserActionTypes.ClearUsers;
}

export class SelectUser implements Action {
  readonly type = UserActionTypes.SelectUser;

  constructor(public payload: { id: number }) {}
}

export type UserActions =
  | LoadUser
  | LoadUserSuccess
  | LoadUserFail
  | LoadUsers
  | LoadUsersSuccess
  | LoadUsersFail
  | AddUser
  | AddUserSuccess
  | AddUserFail
  | AddUsers
  | AddUsersSuccess
  | AddUsersFail
  | UpdateUser
  | UpdateUserSuccess
  | UpdateUserFail
  | UpdateUsers
  | UpdateUsersSuccess
  | UpdateUsersFail
  | DeleteUser
  | DeleteUserSuccess
  | DeleteUserFail
  | DeleteUsers
  | ClearUsers
  | SelectUser;
