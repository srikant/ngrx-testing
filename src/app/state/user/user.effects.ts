import { Injectable } from '@angular/core';
import { UserService } from '@core/services/user.service';
import { Actions, Effect } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable, of } from 'rxjs';
import { catchError, exhaustMap, map } from 'rxjs/operators';
import {
  AddUser,
  AddUserFail,
  AddUserSuccess,
  LoadUser,
  LoadUserFail,
  LoadUserSuccess,
  LoadUsers,
  LoadUsersFail,
  LoadUsersSuccess,
  UpdateUser,
  UpdateUserFail,
  UpdateUserSuccess,
  UserActionTypes
} from './user.actions';

@Injectable()
export class UserEffects {
  @Effect()
  addUser: Observable<Action> = this.actions$
    .ofType<AddUser>(UserActionTypes.AddUser)
    .pipe(
      map(action => action.payload),
      exhaustMap(payload => this.userService.addUser(payload.user)),
      map(user => new AddUserSuccess({ user })),
      catchError(error => of(new AddUserFail({ error })))
    );

  @Effect()
  loadUser: Observable<Action> = this.actions$
    .ofType<LoadUser>(UserActionTypes.LoadUser)
    .pipe(
      map(action => action.payload),
      exhaustMap(payload => this.userService.getUser(payload.id)),
      map(user => new LoadUserSuccess({ user })),
      catchError(error => of(new LoadUserFail({ error })))
    );

  @Effect()
  loadUsers: Observable<Action> = this.actions$
    .ofType<LoadUsers>(UserActionTypes.LoadUsers)
    .pipe(
      exhaustMap(() => this.userService.getUsers()),
      map(users => new LoadUsersSuccess({ users })),
      catchError(error => of(new LoadUsersFail({ error })))
    );

  @Effect()
  updateUser: Observable<Action> = this.actions$
    .ofType<UpdateUser>(UserActionTypes.UpdateUser)
    .pipe(
      map(action => action.payload),
      exhaustMap(payload => this.userService.updateUser(payload.user)),
      map(
        user =>
          new UpdateUserSuccess({
            update: {
              id: user.id,
              changes: user
            }
          })
      ),
      catchError(error => of(new UpdateUserFail({ error })))
    );

  constructor(private actions$: Actions, private userService: UserService) {}
}
