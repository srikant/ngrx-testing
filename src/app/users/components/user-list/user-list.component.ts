import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { User } from '@state/user/user.model';

@Component({
  selector: 'app-user-list',
  templateUrl: './user-list.component.html',
  styleUrls: ['./user-list.component.scss']
})
export class UserListComponent implements OnInit {
  @Input() users: User[];
  @Output() selectUser = new EventEmitter<User>();

  constructor() {}

  ngOnInit() {}
}
