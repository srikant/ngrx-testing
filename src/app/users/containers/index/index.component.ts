import { Component, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { Store, select } from '@ngrx/store';
import { State } from '@state/index';
import { selectAllUsers } from '@state/user';
import { LoadUsers } from '@state/user/user.actions';
import { User } from '@state/user/user.model';
import { Observable } from 'rxjs';

@Component({
  templateUrl: './index.component.html',
  styleUrls: ['./index.component.scss']
})
export class IndexComponent implements OnInit {
  users: Observable<Array<User>>;

  constructor(private router: Router, private store: Store<State>) {}

  ngOnInit() {
    this.store.dispatch(new LoadUsers());
    this.users = this.store.pipe(select(selectAllUsers));
  }

  onSelectUser(user: User) {
    this.router.navigate(['/users', user.id]);
  }
}
