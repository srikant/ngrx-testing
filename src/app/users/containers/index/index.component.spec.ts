import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { Store } from '@ngrx/store';
import { LoadUsers } from '@state/user/user.actions';
import { generateUsers } from '@state/user/user.model';
import { cold, getTestScheduler, hot } from 'jasmine-marbles';
import { last } from 'rxjs/operators';
import { UserListComponent } from './../../components/user-list/user-list.component';
import { IndexComponent } from './index.component';

describe('IndexComponent', () => {
  let component: IndexComponent;
  let fixture: ComponentFixture<IndexComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [IndexComponent, UserListComponent],
      imports: [RouterTestingModule],
      providers: [
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn()
          }
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(IndexComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  describe('ngOnInit()', () => {
    it('should dispatch an the LoadUsers action in ngOnInit lifecycle', () => {
      const action = new LoadUsers();
      const store = TestBed.get(Store);
      const spy = jest.spyOn(store, 'dispatch');

      fixture.detectChanges();

      expect(spy).toHaveBeenCalledWith(action);
    });

    it('should selectAllUsers', () => {
      const store = TestBed.get(Store);
      const users = generateUsers();
      store.pipe = jest.fn(() => hot('-a', { a: users }));

      fixture.detectChanges();

      const expected = cold('-a', { a: users });
      expect(component.users).toBeObservable(expected);
    });
  });

  describe('users', () => {
    it('should be an observable of an array of user objects', done => {
      const users = generateUsers();
      const store = TestBed.get(Store);
      store.pipe = jest.fn(() => cold('-a|', { a: users }));

      fixture.detectChanges();

      component.users.pipe(last()).subscribe(componentUsers => {
        expect(componentUsers).toEqual(users);
        done();
      });

      getTestScheduler().flush();
    });
  });
});
