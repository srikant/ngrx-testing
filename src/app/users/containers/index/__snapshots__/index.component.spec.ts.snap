// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexComponent should match snapshot 1`] = `
<ng-component
  router={[Function Router]}
  store={[Function Object]}
  users="undefined"
>
  <button
    ng-reflect-router-link="add"
    routerlink="add"
    tabindex="0"
  >
    Add User
  </button><app-user-list>
    <ul>
      
    </ul>
  </app-user-list>
</ng-component>
`;
